Add `requires_authorization` Payment state

andrewpbrett opened the pull request:
#### What? Why?

Closes #7687 
Setup for fixing #7665 

This adds a state for payments that require additional authorization from the customer. Previously we were relying on the presence of the `cvv_response_message` field and leaving the payment in the `pending` state. 

This will make it clear that it's a distinct state and is a first step toward clarifying the order + payment state machines. 

<!-- Explain why this change is needed and the solution you propose.
Provide context for others to understand it. -->



#### What should we test?
<!-- List which features should be tested and how. -->
There are changes to the test suite that we should pay close attention to in Code Review and that give some places that could also be tested manually. Everything should work as before related to payments, in particular adding a payment that requires SCA authorization. 


#### Release notes
<!-- Write a one liner description of the change to be included in the release notes.
Every PR is worth mentioning, because you did it for a reason. -->
Updated payments to have an additional state when SCA authorization is required. 
<!-- Please select one for your PR and delete the other. -->
Changelog Category: Technical changes



#### Dependencies
<!-- Does this PR depend on another one?
Add the link or remove this section. -->



#### Documentation updates
<!-- Are there any wiki pages that need updating after merging this PR?
List them here or remove this section. -->

andrewpbrett commented on app/models/spree/payment/processing.rb:
```diff
@@ -240,6 +236,9 @@ def handle_response(response, success_state, failure_state)
             if response.cvv_result
               self.cvv_response_code = response.cvv_result['code']
               self.cvv_response_message = response.cvv_result['message']
+              if self.cvv_response_message.present?
+                return self.require_authorization!
+              end
```
This is the main place that this PR actually changes logic. This is the first one. We now move the payment to the `requires_authorization` state instead of the `pending` state (even though `pend` is passed as the `success_state`. The response comes back as a success, but the state that the payment should end up in depends on whether the response indicates that authorization is required. 

The only case where we're using the `cvv_response_message` field (as far as I can tell) is to store the authorization URL from Stripe.

andrewpbrett commented on app/services/process_payment_intent.rb:
```diff
@@ -36,10 +36,12 @@ def call!
 
     if payment.reload.completed?
-      payment.mark_as_processed
+      payment.completed_authorization
+      payment.clear_authorization_url
```
I kept these as two separate methods for now; `clear_authorization_url` could potentially be an `after_transition` on `completed_authorization`.

Matt-Yorkley replied:
I think I would move it there. We want it to happen whenever that transition completes :+1:

andrewpbrett commented on engines/order_management/app/services/order_management/order/stripe_sca_payment_authorize.rb:
```diff
@@ -15,7 +15,9 @@ def call!(redirect_url = full_order_path(@order))
         @payment.authorize!(redirect_url)
 
-        @order.errors.add(:base, I18n.t('authorization_failure')) unless @payment.pending?
+        unless @payment.pending? || @payment.requires_authorization?
+          @order.errors.add(:base, I18n.t('authorization_failure'))
+        end
```
Because payments requiring auth now move to `requires_authorization` instead of `pending`, this change is necessary to preserve the current behavior.

Matt-Yorkley replied:
Is this needed? Won't line 10 already find only payments that are in `pending` state (which means they're not in `requires_authorization` state)?

andrewpbrett replied:
Line 16 (`@payment.authorize!(redirect_url)`) should either change the state to `requires_authorization` or leave it in `pending`. If there's a problem with the card (insufficient funds for example) it will move it to `failed` and we want to add this error.

luisramos0 replied:
I see "@payment.pending? || @payment.requires_authorization?" appearing twice in this PR, looks like a meaningful condition. I'd make a method in the payment object and explain what it is in the method name or method docs.

andrewpbrett commented on spec/models/spree/payment_spec.rb:
```diff
@@ -27,7 +27,7 @@
                                  authorization: '123',
                                  avs_result: { 'code' => 'avs-code' },
-                                 cvv_result: { 'code' => 'cvv-code', 'message' => "CVV Result" })
+                                 cvv_result: nil)
```
A `nil` `cvv_result` indicates that additional authorization is not required.

Matt-Yorkley replied:
I think it might be `cvv_result: { code: nil, message: nil }` in that case (eg: the `code` and `message` keys are still present, but with `nil` values).

Matt-Yorkley commented on app/models/spree/payment.rb:
```diff
@@ -145,7 +155,7 @@ def adjustment_label
     end
 
-    def mark_as_processed
+    def clear_authorization_url
```
:+1:

luisramos0 replied:
nice! we could add a specific field in the db for this (that was my original idea at least).
payment.authorization_url

Matt-Yorkley commented on app/services/process_payment_intent.rb:
```diff
@@ -36,10 +36,12 @@ def call!
 
       Result.new(ok: true)
     else
+      payment.failed_authorization
```
This will transition it to `failed` state? The payment object will behave differently after we do this, right? Is this what we want?

andrewpbrett replied:
I think it is - if the customer denies authorization on the extra step, we'd treat the payment the same as if it failed for any other reason. They'd have to create a new payment in order to pay the balance on the order.

Matt-Yorkley reviewed (commented):
A couple of thoughts:
- I think there's translation keys for each payment state for where they get shown in the UI, we'll need to add a new key alongside the others. If we want to go nuts on UI clarity; I think the different payment states each have their own CSS colour, and we could add one for this state that's grey or something so it's more clearly distinguished from the others.
- Are there any other places where there's code that should be aware of this new state, or react differently to payments in this state?
- We might want to update `Order::Updater#infer_payment_state` around line 170 and add a check to see if the only payments the order has are in `require_authorization`, and set order.payment_state to "requires_authorization". The mapping of payment states to the order's "payment state" is not 1:1, it's a bit weird but it essentially handles the fact that the order can have multiple payments in different states.

andrewpbrett commented:
- Added translation keys for the new state
- > Are there any other places where there's code that should be aware of this new state, or react differently to payments in this state?
Yes, I think this should help when we're creating admin payments for example - if we do an `authorize` on a payment we can then check if its state is `requires_authorization` and not try to process the payment in that case (since calling `capture` will fail). 
- Added logic to the order updater to address this

I think the question is whether we go with something small, like this, or try to add the solution #7665 (and possibly other improvements to the way we handle payment states in other places) in the same PR.

Matt-Yorkley commented on app/models/spree/order/checkout.rb:
```diff
@@ -67,12 +67,6 @@ def self.define_state_machine!
 
-              if states[:payment]
-                before_transition to: :complete do |order|
-                  order.process_payments! if order.payment_required?
```
I'm not sure we can remove this..?

andrewpbrett replied:
Yeah, it was definitely a "let's take this out and see what breaks" commit. I'm going to look through today and see where we'd need to change things and if it's manageable. It would be really nice to separate the two things and remove a callback.

Matt-Yorkley replied:
Decoupling payment processing from order completion might end up being a pretty big refactor. Maybe it's worth doing it in a separate PR?

Matt-Yorkley commented:
Alright, there's a couple of build errors that need sorting out but this is looking great :tada:

andrewpbrett commented:
> Decoupling payment processing from order completion might end up being a pretty big refactor. Maybe it's worth doing it in a separate PR?

Good idea @Matt-Yorkley, thanks for saving me from going off on a tangent :-) I think this PR can stand on its own and I moved the last few commits to a separate branch, which now only has a handful of specs to fix: https://github.com/andrewpbrett/openfoodnetwork/runs/2810511792

Matt-Yorkley commented on spec/features/admin/payments_stripe_spec.rb:
```diff
@@ -82,7 +82,7 @@
 
           expect(page).to have_link "StripeSCA"
           expect(page).to have_content "PENDING"
```
This test looks a bit weird now. Should it say "pending" in the UI here?

Also the test description is "fails to add a payment due to card error", maybe that needs updating?

andrewpbrett replied:
Updated 👍 ccc873fd25eaa42be8875556c4f1195318843296

Matt-Yorkley commented on app/models/spree/credit_card.rb:
```diff
@@ -83,12 +83,12 @@ def actions
 
     def can_resend_authorization_email?(payment)
-      payment.pending? && payment.authorization_action_required?
+      payment.pending? && payment.requires_authorization?
```
This needs updating, right? It's one state or the other, but not both?

andrewpbrett replied:
Yes, that's right. Good catch 👍

jaycmb commented:
Will add the requires_authorization state to the [Order States Flow Chart](https://miro.com/welcomeonboard/SkJ6c1BiRlNrd2o0aXQwUmpmWXNyVHZvU1Y2YkRaaTN2RE5qYTh1ZE14Z0RibFNIYnZyM1dWN3JDQXl0UnlXR3wzMDc0NDU3MzQ2NjQ5NTUzODI1) 
Ping @Filipe.

And doublecheck: this came out of the Admin order PR [#7709 ](https://github.com/openfoodfoundation/openfoodnetwork/pull/7709#event-4874863572), but currently Admin order flow does not have state `pending` but `none` when Authentication fails

<img width="907" alt="image" src="https://user-images.githubusercontent.com/39744592/121877850-5b001b00-cd0b-11eb-8ede-27eb57196704.png">

And will this intermediate states then apply for all flows  (sub, admin orders and standard shopfront process), right @andrewpbrett ?

andrewpbrett commented:
Thanks @jaycmb!

> currently Admin order flow does not have state pending but none when Authentication fails

Yes, after this PR, when authentication fails, the payment should move to the `failed` state rather than no state (current behavior)

> And will this intermediate states then apply for all flows

Yes again :) After this PR, all three types of payments should move to `requires_authorization` if necessary. This could potentially be very brief in the case of a shopfront payment.

luisramos0 reviewed (approved):
:clap: very good PR.

filipefurtad0 commented:
Hey @andrewpbrett ,

Thanks for this PR! I think this new payment state will add some clarity in our order-states flow. **Shopfront checkout** was tested, with a card which always requires authentication (x-3184), this seems to be breaking now:

![Peek 2021-06-16 16-07](https://user-images.githubusercontent.com/49817236/122246781-c2aa9780-cebe-11eb-9766-d26e947b5dc8.gif)

I monitored the changes on the respective payment, to "catch" the transition from/to `requires_authoriaztion` state (left side). All good in there, the payment reaches this state, once authentication is requested. After granting it the customer does not get redirected to the order confirmation page, and payments are not captured. However, the **order does reach a complete state**:

```
   id    |   number   | item_total | total |  state   | adjustment_total | user_id |         created_at         |         updated_at         |        completed_at        | bill_address_id | ship_address_id | payment_total | shipment_state |     payment_state      |               email               | special_instructions | distributor_id | currency | last_ip_address | order_cycle_id | customer_id | created_by_id | included_tax_total | additional_tax_total 
---------+------------+------------+-------+----------+------------------+---------+----------------------------+----------------------------+----------------------------+-----------------+-----------------+---------------+----------------+------------------------+-----------------------------------+----------------------+----------------+----------+-----------------+----------------+-------------+---------------+--------------------+----------------------
 1981416 | R202127673 |       6.00 |  6.00 | complete |             0.00 |   19499 | 2021-06-16 07:56:36.957452 | 2021-06-16 15:09:05.784945 | 2021-06-16 15:09:05.109136 |          782901 |          782902 |          0.00 | pending        | requires_authorization | filipefurtado+baguettec@gmail.com |                      |           2489 | EUR      | 176.79.33.88    |           7295 |       21184 |         19499 |               1.00 |                 0.00
(1 row)
```

It's also seen in the BO:

![image](https://user-images.githubusercontent.com/49817236/122247588-5e3c0800-cebf-11eb-9da9-c4e924dc5e8d.png)

**Backoffice orders**

Creating a new order, and after adding an SCA payment with the x-3184 card does not lead to payment capture either:

![image](https://user-images.githubusercontent.com/49817236/122247739-7d3a9a00-cebf-11eb-9a13-2f5d74110ac4.png)

In this case the order remains in the `payment` state 

```
   id    |   number   | item_total | total |  state  | adjustment_total | user_id |         created_at         |        updated_at         | completed_at | bill_address_id | ship_address_id | payment_total | shipment_state | payment_state |               email               | special_instructions | distributor_id | currency | last_ip_address | order_cycle_id | customer_id | created_by_id | included_tax_total | additional_tax_total 
---------+------------+------------+-------+---------+------------------+---------+----------------------------+---------------------------+--------------+-----------------+-----------------+---------------+----------------+---------------+-----------------------------------+----------------------+----------------+----------+-----------------+----------------+-------------+---------------+--------------------+----------------------
 1981419 | R521831342 |       7.50 |  7.50 | payment |             0.00 |   19499 | 2021-06-16 15:10:19.334044 | 2021-06-16 15:11:16.44216 |              |          782903 |          782904 |          0.00 |                |               | filipefurtado+baguettec@gmail.com |                      |           2489 | EUR      |                 |           7295 |       21184 |               |               1.25 |                 0.00
(1 row)
```

**Subscription orders**

Order placement fails, as indicated per email,

![image](https://user-images.githubusercontent.com/49817236/122250623-b83dcd00-cec1-11eb-833c-c69cdb7869dd.png)

but the order transitions from `payment` to `complete` state:

```
   id    |   number   | item_total | total |  state   | adjustment_total | user_id |         created_at         |         updated_at         |        completed_at        | bill_address_id | ship_address_id | payment_total | shipment_state |     payment_state      |               email               | special_instructions | distributor_id | currency | last_ip_address | order_cycle_id | customer_id | created_by_id | included_tax_total | additional_tax_total 
---------+------------+------------+-------+----------+------------------+---------+----------------------------+----------------------------+----------------------------+-----------------+-----------------+---------------+----------------+------------------------+-----------------------------------+----------------------+----------------+----------+-----------------+----------------+-------------+---------------+--------------------+----------------------
 1981420 | R855875244 |      66.66 | 80.66 | complete |            14.00 |   19499 | 2021-06-16 15:35:31.483363 | 2021-06-16 15:36:11.966585 | 2021-06-16 15:35:33.863488 |          782905 |          782906 |          0.00 | pending        | requires_authorization | filipefurtado+baguettec@gmail.com |                      |           2489 | EUR      |                 |           7291 |       21184 |               |               6.58 |                 0.00
(1 row)
```

Using the link to provide payment authentication leads to the same outcome as before:

![image](https://user-images.githubusercontent.com/49817236/122250907-f9ce7800-cec1-11eb-8ccd-3d03380e5bd1.png)

I'm moving back to In Dev for now, please let me know if there is any additional scenario I should check or if you need additional feedback :+1:

andrewpbrett commented:
Thanks @filipefurtad0 - I think this might turn out to be tied too closely to #7781 and we may have to test with that PR instead of teasing them apart. I was able to reproduce your test result and I'm able to do the checkout on the #7781 branch. I'll test the subscriptions and BO scenarios next... but I might close this one or pull in the commits from #7781 to this branch instead.

andrewpbrett commented:
Okay, I was actually able to tease these apart still. With that latest push I tested BO and shopfront orders with the SCA-required card number and they both seemed to work. I think we can put this into testing again.

filipefurtad0 commented:
Hey @andrewpbrett , this PR has some conflicts - could you have a look? Thanks :+1:

sauloperez commented:
Solved @filipefurtad0

filipefurtad0 commented:
Hey @andrewpbrett ,

Staged the PR and proceed as before, tracking state-changes vs. actions:


**1) Shopfront orders**

The after adding the customer details, adding a Stripe SCA (x-3184) payment and clicking "Place Order" changes the order from a `cart` state to a `payment` state:

```

   id    |   number   | item_total | total |  state  | adjustment_total | user_id |         created_at         |        updated_at         | completed_at | bill_address_id | ship_address_id | payment_total | shipment_state | payment_state |               email               | special_instructions | distributor_id | currency | last_ip_address | order_cycle_id | customer_id | created_by_id | included_tax_total | additional_tax_total 
---------+------------+------------+-------+---------+------------------+---------+----------------------------+---------------------------+--------------+-----------------+-----------------+---------------+----------------+---------------+-----------------------------------+----------------------+----------------+----------+-----------------+----------------+-------------+---------------+--------------------+----------------------
 1981428 | R117466277 |       4.50 |  4.50 | payment |             0.00 |   19499 | 2021-06-17 09:57:31.778187 | 2021-06-18 10:55:28.51834 |              |          782923 |          782924 |          0.00 |                |               | filipefurtado+baguettec@gmail.com |                      |           2489 | EUR      | 37.189.1.149    |           7295 |       21184 |         19499 |               0.75 |                 0.00
(1 row)
```

Completing authentication (Stripe test page) further changes it to `complete`:

```
   id    |   number   | item_total | total |  state   | adjustment_total | user_id |         created_at         |         updated_at         |        completed_at        | bill_address_id | ship_address_id | payment_total | shipment_state | payment_state |               email               | special_instructions | distributor_id | currency | last_ip_address | order_cycle_id | customer_id | created_by_id | included_tax_total | additional_tax_total 
---------+------------+------------+-------+----------+------------------+---------+----------------------------+----------------------------+----------------------------+-----------------+-----------------+---------------+----------------+---------------+-----------------------------------+----------------------+----------------+----------+-----------------+----------------+-------------+---------------+--------------------+----------------------
 1981428 | R117466277 |       4.50 |  4.50 | complete |             0.00 |   19499 | 2021-06-17 09:57:31.778187 | 2021-06-18 10:58:03.640888 | 2021-06-18 10:58:03.540554 |          782453 |          782454 |          4.50 | ready          | paid          | filipefurtado+baguettec@gmail.com |                      |           2489 | EUR      | 37.189.1.149    |           7295 |       21184 |         19499 |               0.75 |                 0.00
(1 row)
```

On the payments side, from the point where it is created ("Place order", on `/checkout` page) payment has the new payment state `requires_authorization` introduced in this PR:

```

   id   | amount | order_id |         created_at         |         updated_at         | source_id |    source_type    | payment_method_id |         state          |        response_code        | avs_response | identifier | cvv_response_code |                                                                                                                cvv_response_message                                                                                                                 
--------+--------+----------+----------------------------+----------------------------+-----------+-------------------+-------------------+------------------------+-----------------------------+--------------+------------+-------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 159930 |   4.50 |  1981428 | 2021-06-18 10:55:25.439364 | 2021-06-18 10:55:28.450512 |     22664 | Spree::CreditCard |               629 | requires_authorization | pi_1J3fMBKuuB1fWySn2MAIfp30 |              | TEK7JAHC   |                   | https://hooks.stripe.com/redirect/authenticate/src_1J3fMCKuuB1fWySn3xpnxWEc?client_secret=src_client_secret_Nu7XHhByT4fCbXT2oiPejzvt&source_redirect_slug=test_YWNjdF8xRmlxRXNLdXVCMWZXeVNuLF9KaDNUdmFSbE82SEhmTWxPaEtVVkJZRTZjdzRoZHRN0100lC1UQaXN
(1 row)
```

Completing authentication (Stripe test page) further changes it to `completed`:

```
   id   | amount | order_id |         created_at         |         updated_at         | source_id |    source_type    | payment_method_id |   state   |        response_code        | avs_response | identifier | cvv_response_code | cvv_response_message 
--------+--------+----------+----------------------------+----------------------------+-----------+-------------------+-------------------+-----------+-----------------------------+--------------+------------+-------------------+----------------------
 159930 |   4.50 |  1981428 | 2021-06-18 10:55:25.439364 | 2021-06-18 10:58:03.449449 |     22664 | Spree::CreditCard |               629 | completed | pi_1J3fMBKuuB1fWySn2MAIfp30 |              | TEK7JAHC   |                   | 
(1 row)
```

So, checkout works now :tada: 

**2) Backoffice orders**

Following the same procedure, and order in the BO starts as `cart` and goes to `payment`, once adding customer details are added. It stays in that state, even when a Stripe SCA payment (x-3184) is created.

On the payments side, it works very similarly as to what is described above for shopfront orders: the payment starts off with the state `requires_authorization` - this is correctly reflected in the UI:

![image](https://user-images.githubusercontent.com/49817236/122553925-34562300-d030-11eb-9870-cdb46f4eca3a.png)

Once the customer provides authorization - link received via Email, or under `/account` - the order goes to the `complete` state, and the payment goes from `requires_authorization `to `completed`

![image](https://user-images.githubusercontent.com/49817236/122554767-4e443580-d031-11eb-9bd0-2b267fdcb645.png)

**3) Subscription orders**

Even before the OC opens, the proxy order creates an order in the state `complete`, with an associated payment, on state `checkout`:

![image](https://user-images.githubusercontent.com/49817236/122555487-42a53e80-d032-11eb-9ca5-03ba7691ef8f.png)

After running the placement job (the OC opens):
`bundle exec rake ofn:subs:test:repeat_placement_job`

At this point, nothing changes on what states is concerned. However, after running the confirmation job:
`bundle exec rake ofn:subs:test:force_confirmation_job`

the payment state changes from `checkout` to the expected `requires_authorization`:

![image](https://user-images.githubusercontent.com/49817236/122556186-1a6a0f80-d033-11eb-9ab4-258f8afadea3.png)

After the customer authorizes and the authentication step (Stripe) is completed:

![image](https://user-images.githubusercontent.com/49817236/122556622-a54b0a00-d033-11eb-930e-1e9cecea07d4.png)

I'd say this ready to go, the new state is introduced right where it is supposed to, to my understanding of the flow and orders states chart. Ping @jaycmb .

Awesome work team :clap: :clap: :clap: 

I'm moving to ready to go, not to hold the preparing of the release any longer.

sauloperez commented:
This brings clarity :clap:
